From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Mon, 17 Jul 2023 20:11:44 +0200
Subject: [PATCH] Resend more data on locale change

Includes advancements, inventories, entity custom names, text display texts and item frame items

diff --git a/net/minecraft/server/PlayerAdvancements.java b/net/minecraft/server/PlayerAdvancements.java
index 98004b77b7f40d315f41be82105df6d4b9000e33..b8562c204900e646f51c75ed5069656b21af6d43 100644
--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -71,6 +71,16 @@ public class PlayerAdvancements {
         this.load(manager);
     }
 
+    // CloudPlane start - refresh data on locale change; advancement refreshing
+    public void refreshAll() {
+        this.isFirstPacket = true;
+        for (AdvancementHolder advancement : this.visible) {
+            this.markForVisibilityUpdate(advancement);
+        }
+        this.visible.clear();
+    }
+    // CloudPlane end - refresh data on locale change; advancement refreshing
+
     public void setPlayer(ServerPlayer player) {
         this.player = player;
     }
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 79c804245c7f41a5a7f062e1c5e01760adde7d84..a194541fc5c0cbf1384fce09e90766ec3e50b5c2 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -464,6 +464,8 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     }
     // Paper end - rewrite chunk system
 
+    private long lastLocaleChange = 0L; // CloudPlane - refresh data on locale change; throttling
+
     public ServerPlayer(MinecraftServer server, ServerLevel level, GameProfile gameProfile, ClientInformation clientInformation) {
         super(level, gameProfile);
         this.server = server;
@@ -2427,6 +2429,33 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                 clientInformation.language()
             );
             this.server.server.getPluginManager().callEvent(event);
+
+            // CloudPlane start - refresh data on locale change
+            if (this.lastLocaleChange + 1000L < System.currentTimeMillis()) {
+                this.lastLocaleChange = System.currentTimeMillis();
+                this.advancements.refreshAll();
+                this.inventoryMenu.sendAllDataToRemote();
+                if (this.containerMenu != this.inventoryMenu) {
+                    this.containerMenu.sendAllDataToRemote();
+                }
+
+                for (Entity entity : this.level().getEntities().getAll()) {
+                    if (!entity.moonrise$getTrackedEntity().seenBy.contains(this.connection)) {
+                        continue;
+                    }
+
+                    if (entity.hasCustomName()) {
+                        entity.getEntityData().markDirty(Entity.DATA_CUSTOM_NAME);
+                    }
+                    if (entity instanceof net.minecraft.world.entity.Display.TextDisplay) {
+                        entity.getEntityData().markDirty(net.minecraft.world.entity.Display.TextDisplay.DATA_TEXT_ID);
+                    } else if (entity instanceof net.minecraft.world.entity.decoration.ItemFrame
+                            && !((net.minecraft.world.entity.decoration.ItemFrame) entity).getItem().isEmpty()) {
+                        entity.getEntityData().markDirty(net.minecraft.world.entity.decoration.ItemFrame.DATA_ITEM);
+                    }
+                }
+            }
+            // CloudPlane end - refresh data on locale change
         }
         // CraftBukkit end
         // Paper start - don't call options events on login
